linux-pipewire: Support enumerated camera sizes

pobrn opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description + Motivation and Context

A node might choose to use an enumerated choice pod to report the set of
supported sizes. Currently that is not supported. And what's worse, since
`SPA_POD_OPT_Rectangle()` is used, if the node uses an enum choice, then
`{,this_}resolution` stays uninitialized but `spa_pod_parse_object()` will
not report errors because the field is considered optional.

Fix this by using `spa_pod_get_values()`, which can be used for concrete
values, and both `SPA_CHOICE_{None,Enum}`.

Since the resolutions are parsed in two separate places, and since the
frame rates require essentially the same treatment a new function is
introduced to handle the common parts.

### How Has This Been Tested?

With the current pipewire master branch, with a video node that uses enumerated choice for size, as well as nodes that don't use enumerated choices.


### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
- Bug fix (non-breaking change which fixes an issue)
<!--- - New feature (non-breaking change which adds functionality) -->
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

PatTheMav commented on plugins/linux-pipewire/camera-portal.c:
```diff
@@ -499,19 +499,66 @@ static bool device_selected(void *data, obs_properties_t *props, obs_property_t
 }
 
-static int sort_resolutions(gconstpointer a, gconstpointer b)
+static void remove_duplicates_from_sorted(GArray *array, GCompareFunc cmp)
```
Is #11741 required for this PR then?

RytoEX replied:
It sounds more like #11741 supersedes this PR.

pobrn committed:
linux-pipewire: Simplify camera frame rate extraction

Add a new function `get_values_from_pod()` to provide a wrapper around
`spa_pod_get_values()` with extra checks, and that ignores the default
value for `SPA_CHOICE_Enum`.

pobrn committed:
linux-pipewire: Support enumerated camera frame sizes

A node might choose to use an enumerated choice pod to report the set of
supported sizes. Currently that is not supported. And what's worse, since
`SPA_POD_OPT_Rectangle()` is used, if the node uses an enum choice, then
`{,this_}resolution` stays uninitialized but `spa_pod_parse_object()` will
not report errors because the field is considered optional.

Fix this by handling `SPA_CHOICE_Enum` as well, and ignoring the unsupported
choice types.

jcm93 commented:
For what it's worth, the choice I made in #10895 was also to omit a parameter in the format string if that parameter was empty or otherwise meaningless (and a format composed of all-meaningless values would be coalesced to the string "Default"). On macOS this was very rare, but could be the case for devices like wired iPhone screen sharing.

pobrn commented:
Is there any way this can progress? I see #12781 addresses the same thing as the last commit here.

tytan652 commented:
#12781 is a bug fix for a specific "regression" that appeared OBS 32 meant to be merge for a patch release without adding features.

This PR is a feature addition.
